# <font color = 'lightblue'> TRABAJO EDA APRENDIZAJE AUTOMÁTICO


**Guzmán Muñoz Revuelta**

**Fernando Caruana del Pino**

**Importamos las librerías necesarias**

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
from matplotlib import pyplot as plt
import plotly.express as px

pd.set_option('display.max_columns', 500) #esto es para que me muestre todas las columnas
pd.set_option('display.max_rows', 5000) #esto es para que me muestre todas las filas

In [2]:
bf = pd.read_csv('../data/Base.csv', sep=',', decimal='.')
bf.head() #observamos las 5 primeras filas del dataset

fraud_bool  income  name_email_similarity  prev_address_months_count  \
0           1     0.9               0.166828                         -1   
1           1     0.9               0.296286                         -1   
2           1     0.9               0.044985                         -1   
3           1     0.9               0.159511                         -1   
4           1     0.9               0.596414                         -1   

   current_address_months_count  customer_age  days_since_request  \
0                            88            50            0.020925   
1                           144            50            0.005418   
2                           132            40            3.108549   
3                            22            50            0.019079   
4                           218            50            0.004441   

   intended_balcon_amount payment_type  zip_count_4w   velocity_6h  \
0               -1.331345           AA           769  10650.765523   
1               -0.816224           AB           366    534.047319   
2               -0.755728           AC           870   4048.534263   
3               -1.205124           AB           810   3457.064063   
4               -0.773276           AB           890   5020.341679   

   velocity_24h  velocity_4w  bank_branch_count_8w  \
0   3134.319630  3863.647740                     1   
1   2670.918292  3124.298166                   718   
2   2893.621498  3159.590679                     1   
3   4054.908412  3022.261812                  1921   
4   2728.237159  3087.670952                  1990   

   date_of_birth_distinct_emails_4w employment_status  credit_risk_score  \
0                                 6                CA                185   
1                                 3                CA                259   
2                                14                CB                177   
3                                 6                CA                110   
4                                 2                CA                295   

   email_is_free housing_status  phone_home_valid  phone_mobile_valid  \
0              0             BA                 1                   0   
1              1             BA                 0                   0   
2              1             BA                 0                   1   
3              1             BA                 0                   1   
4              1             BA                 1                   0   

   bank_months_count  has_other_cards  proposed_credit_limit  foreign_request  \
0                 24                0                  500.0                0   
1                 15                0                 1500.0                0   
2                 -1                0                  200.0                0   
3                 31                1                  200.0                0   
4                 31                0                 1500.0                0   

     source  session_length_in_minutes  device_os  keep_alive_session  \
0  INTERNET                   3.888115    windows                   0   
1  INTERNET                  31.798819    windows                   0   
2  INTERNET                   4.728705      other                   0   
3  INTERNET                   2.047904      linux                   0   
4  INTERNET                   3.775225  macintosh                   1   

   device_distinct_emails_8w  device_fraud_count  month  
0                          1                   0      7  
1                          1                   0      7  
2                          1                   0      7  
3                          1                   0      7  
4                          1                   0      7

# **OBJETIVOS DEL TRABAJO**

El objetivo de este trabajo es abordar el desafío de la detección de fraudes bancarios utilizando un conjunto de datos de detección de fraudes bancarios. La variable principal en este estudio es "bf" ("bank fraud") que representa la presencia o ausencia de fraude en las transacciones bancarias. La detección de fraudes bancarios es crucial para las instituciones financieras y los clientes con el fin de prevenir pérdidas económicas y mantener la confianza en el sistema bancario.

El trabajo se enfoca en los siguientes puntos clave:

1. Conjunto de Datos: Se utiliza un conjunto de datos que contiene información detallada sobre transacciones bancarias, incluyendo características relevantes como el monto de la transacción, la hora, el tipo de tarjeta utilizada, entre otros.

2. Análisis Exploratorio de Datos (EDA): Se realiza un EDA para comprender la distribución de la variable "bf" y explorar las relaciones entre las diferentes características y la variable objetivo. Esto ayuda a identificar posibles patrones de fraude.

3. Preprocesamiento de Datos: Se lleva a cabo un preprocesamiento de datos que puede incluir la normalización de características, la eliminación de valores atípicos y la codificación de variables categóricas.

4. Selección de Modelo: Se evalúan varios modelos de aprendizaje automático, como Regresión Logística, Random Forest, Support Vector Machines, o Redes Neuronales, para determinar cuál es el más adecuado para la detección de fraudes bancarios.

5. Entrenamiento y Evaluación del Modelo: El modelo seleccionado se entrena utilizando una parte del conjunto de datos y se evalúa utilizando métricas de rendimiento, como la precisión, la sensibilidad y la especificidad.

6. Optimización del Modelo: Se pueden realizar ajustes en el modelo y los hiperparámetros para mejorar su rendimiento.

7. Implementación en Producción: Una vez que se obtiene un modelo de detección de fraudes eficaz, se puede implementar en la operación bancaria en tiempo real para identificar transacciones sospechosas.

Este trabajo tiene como objetivo proporcionar una solución efectiva para la detección de fraudes bancarios utilizando la variable "bf" como variable principal. La detección de fraudes es esencial en el entorno financiero para prevenir pérdidas y proteger a los clientes de posibles actividades fraudulentas.

# Variables

• income (numeric): Annual income of the applicant (in decile form). Ranges between [0.1, 0.9].

• name_email_similarity (numeric): Metric of similarity between email and applicant’s name. Higher values represent higher similarity. Ranges between [0, 1].

• prev_address_months_count (numeric): Number of months in previous registered address of the applicant, i.e. the applicant’s previous residence, if applicable. Ranges between [−1, 380] months (-1 is a missing value).

• current_address_months_count (numeric): Months in currently registered address of the applicant. Ranges between [−1, 429] months (-1 is a missing value).

• customer_age (numeric): Applicant’s age in years, rounded to the decade. Ranges between [10, 90] years.

• days_since_request (numeric): Number of days passed since application was done. Ranges between [0, 79] days.

• intended_balcon_amount (numeric): Initial transferred amount for application. Ranges between [−16, 114] (negatives are missing values).

• payment_type (categorical): Credit payment plan type. 5 possible (annonymized) values.

• zip_count_4w (numeric): Number of applications within same zip code in last 4 weeks. Ranges between [1, 6830].

• velocity_6h (numeric): Velocity of total applications made in last 6 hours i.e., average number of applications per hour in the last 6 hours. Ranges between [−175, 16818].

• velocity_24h (numeric): Velocity of total applications made in last 24 hours i.e., average number of applications per hour in the last 24 hours. Ranges between [1297, 9586]

• velocity_4w (numeric): Velocity of total applications made in last 4 weeks, i.e., average number of applications per hour in the last 4 weeks. Ranges between [2825, 7020].

• bank_branch_count_8w (numeric): Number of total applications in the selected bank branch in last 8 weeks. Ranges between [0, 2404].

• date_of_birth_distinct_emails_4w (numeric): Number of emails for applicants with same date of birth in last 4 weeks. Ranges between [0, 39].

• employment_status (categorical): Employment status of the applicant. 7 possible (annonymized) values. 

• credit_risk_score (numeric): Internal score of application risk. Ranges between [−191, 389].

• email_is_free (binary): Domain of application email (either free or paid).

• housing_status (categorical): Current residential status for applicant. 7 possible (annonymized) values.

• phone_home_valid (binary): Validity of provided home phone.

• phone_mobile_valid (binary): Validity of provided mobile phone.

• bank_months_count (numeric): How old is previous account (if held) in months. Ranges between [−1, 32] months (-1 is a missing value).

• has_other_cards (binary): If applicant has other cards from the same banking company.

• proposed_credit_limit (numeric): Applicant’s proposed credit limit. Ranges between [200, 2000].

• foreign_request (binary): If origin country of request is different from bank’s country.

• source (categorical): Online source of application. Either browser (INTERNET) or app (TELEAPP).

• session_length_in_minutes (numeric): Length of user session in banking website in minutes. Ranges between [−1, 107] minutes (-1 is a missing value).

• device_os (categorical): Operative system of device that made request. Possible values are: Windows, macOS, Linux, X11, or other.

• keep_alive_session (binary): User option on session logout.

• device_distinct_emails (numeric): Number of distinct emails in banking website from the used device in last 8 weeks. Ranges between [−1, 2] emails (-1 is a missing value).

• device_fraud_count (numeric): Number of fraudulent applications with used device. Ranges between [0, 1].

• month (numeric): Month where the application was made. Ranges between [0, 7].

• fraud_bool (binary): If the application is fraudulent or not.

# TRATAMIENTO DE LAS VARIABLES

Las variables originales no están en su tipo de variable correspondiente (algunas) por lo que tenemos que cambiarlas:

In [3]:
bf.dtypes

fraud_bool                            int64
income                              float64
name_email_similarity               float64
prev_address_months_count             int64
current_address_months_count          int64
customer_age                          int64
days_since_request                  float64
intended_balcon_amount              float64
payment_type                         object
zip_count_4w                          int64
velocity_6h                         float64
velocity_24h                        float64
velocity_4w                         float64
bank_branch_count_8w                  int64
date_of_birth_distinct_emails_4w      int64
employment_status                    object
credit_risk_score                     int64
email_is_free                         int64
housing_status                       object
phone_home_valid                      int64
phone_mobile_valid                    int64
bank_months_count                     int64
has_other_cards                 

In [4]:
# creamos un diccionario con las variables, cambiando el tipo de dato a las que nos interesan.
new_types = {'fraud_bool': 'category', 'income': 'float', 'name_email_similarity': 'float', 'prev_address_months_count': 'int64', 'current_address_months_count': 'int64', 'customer_age': 'int64', 'days_since_request': 'float64', 'intended_balcon_amount': 'float', 'payment_type': 'category', 'zip_count_4w': 'int64', 'velocity_6h': 'float', 'velocity_24h': 'float', 'velocity_4w': 'float', 'bank_branch_count_8w': 'int64', 'date_of_birth_distinct_emails_4w': 'int64', 'employment_status': 'category', 'credit_risk_score': 'int64', 'email_is_free': 'category', 'housing_status': 'category', 'phone_home_valid': 'category', 'phone_mobile_valid': 'category', 'bank_months_count': 'int64', 'has_other_cards': 'category', 'proposed_credit_limit': 'int64', 'foreign_request': 'category', 'source': 'category', 'session_length_in_minutes': 'float', 'device_os': 'category', 'keep_alive_session': 'category', 'device_distinct_emails_8w': 'int64', 'device_fraud_count': 'int64', 'month': 'int64'}

# cambiamos esas variables en nuestro dataset
new_bf = bf.astype(new_types)


In [5]:
new_bf.dtypes
#ahora las tenemos como queríamos en un principio.

fraud_bool                          category
income                               float64
name_email_similarity                float64
prev_address_months_count              int64
current_address_months_count           int64
customer_age                           int64
days_since_request                   float64
intended_balcon_amount               float64
payment_type                        category
zip_count_4w                           int64
velocity_6h                          float64
velocity_24h                         float64
velocity_4w                          float64
bank_branch_count_8w                   int64
date_of_birth_distinct_emails_4w       int64
employment_status                   category
credit_risk_score                      int64
email_is_free                       category
housing_status                      category
phone_home_valid                    category
phone_mobile_valid                  category
bank_months_count                      int64
has_other_

In [6]:
bf

fraud_bool  income  name_email_similarity  prev_address_months_count  \
0                1     0.9               0.166828                         -1   
1                1     0.9               0.296286                         -1   
2                1     0.9               0.044985                         -1   
3                1     0.9               0.159511                         -1   
4                1     0.9               0.596414                         -1   
...            ...     ...                    ...                        ...   
999995           0     0.6               0.192631                         -1   
999996           0     0.8               0.322989                        148   
999997           0     0.8               0.879403                         -1   
999998           0     0.9               0.762112                         -1   
999999           0     0.2               0.697452                         -1   

        current_address_months_count  customer_age  days_since_request  \
0                                 88            50            0.020925   
1                                144            50            0.005418   
2                                132            40            3.108549   
3                                 22            50            0.019079   
4                                218            50            0.004441   
...                              ...           ...                 ...   
999995                           104            40            0.030592   
999996                             9            50            1.628119   
999997                            30            20            0.018563   
999998                           189            20            0.015352   
999999                           321            20            2.655916   

        intended_balcon_amount payment_type  zip_count_4w   velocity_6h  \
0                    -1.331345           AA           769  10650.765523   
1                    -0.816224           AB           366    534.047319   
2                    -0.755728           AC           870   4048.534263   
3                    -1.205124           AB           810   3457.064063   
4                    -0.773276           AB           890   5020.341679   
...                        ...          ...           ...           ...   
999995               -1.044454           AB           804   7905.711839   
999996               -1.409803           AC          3306   5391.470463   
999997               34.692760           AA          1522   8063.102636   
999998               94.661055           AA          1418   8092.641762   
999999                9.908499           AA           951   6169.630036   

        velocity_24h  velocity_4w  bank_branch_count_8w  \
0        3134.319630  3863.647740                     1   
1        2670.918292  3124.298166                   718   
2        2893.621498  3159.590679                     1   
3        4054.908412  3022.261812                  1921   
4        2728.237159  3087.670952                  1990   
...              ...          ...                   ...   
999995   8341.468557  4972.635997                     1   
999996   4955.170808  5022.728108                     0   
999997   5670.654316  4377.196321                  2023   
999998   3982.582204  4394.803296                  1678   
999999   3695.308261  4352.334543                     2   

        date_of_birth_distinct_emails_4w employment_status  credit_risk_score  \
0                                      6                CA                185   
1                                      3                CA                259   
2                                     14                CB                177   
3                                      6                CA                110   
4                                      2                CA                295   
...                                  ...               ...    

## Analisis generales de la tabla

Dimensiones

In [7]:
print(bf.shape, bf.drop_duplicates().shape) #en esta base de datos hay 1 millón de filas y 32 columnas

(1000000, 32) (1000000, 32)


**Tipos de datos**

In [8]:
new_bf.dtypes.to_dict() #los tipos de datos

{'fraud_bool': CategoricalDtype(categories=[0, 1], ordered=False),
 'income': dtype('float64'),
 'name_email_similarity': dtype('float64'),
 'prev_address_months_count': dtype('int64'),
 'current_address_months_count': dtype('int64'),
 'customer_age': dtype('int64'),
 'days_since_request': dtype('float64'),
 'intended_balcon_amount': dtype('float64'),
 'payment_type': CategoricalDtype(categories=['AA', 'AB', 'AC', 'AD', 'AE'], ordered=False),
 'zip_count_4w': dtype('int64'),
 'velocity_6h': dtype('float64'),
 'velocity_24h': dtype('float64'),
 'velocity_4w': dtype('float64'),
 'bank_branch_count_8w': dtype('int64'),
 'date_of_birth_distinct_emails_4w': dtype('int64'),
 'employment_status': CategoricalDtype(categories=['CA', 'CB', 'CC', 'CD', 'CE', 'CF', 'CG'], ordered=False),
 'credit_risk_score': dtype('int64'),
 'email_is_free': CategoricalDtype(categories=[0, 1], ordered=False),
 'housing_status': CategoricalDtype(categories=['BA', 'BB', 'BC', 'BD', 'BE', 'BF', 'BG'], ordered=False)

## Exploración de la variable objetivo y tratamiento

In [9]:
bf_fraud_bool = bf['fraud_bool']\
        .value_counts(normalize=True)\
        .mul(100).rename('percent').reset_index()

bf_fraud_bool_conteo = bf['fraud_bool'].value_counts().reset_index()
bf_fraud_bool_pc = pd.merge(bf_fraud_bool, bf_fraud_bool_conteo, on=['index'], how='inner')
bf_fraud_bool_pc

index  percent  fraud_bool
0      0  98.8971      988971
1      1   1.1029       11029

Elegimos nuestra variable objetivo, en este caso es 'fraud_bool', es decir, los fraudes booleanos que existen en nuestro dataframe. Se puede observar una clara diferencia, donde mayoritariamente destaca el valor '0', que equivaldría a no fraude. 

Hemos creado la variable 'bf_fraud_bool_pc', donde combinamos (merge) la variable de fraudes ('bf_fraud_bool') con el conteo de los posibles casos (988971 ceros/no fraudes y 11029 unos/fraudes), para obtener el porcentaje de casos.

In [10]:
fig = px.bar(bf_fraud_bool_pc, x="index", y=['percent'])
fig.update_xaxes(categoryorder='category ascending')
fig.update_layout(xaxis={'categoryorder':'array', 'categoryarray': ['No Fraude', 'Fraude']})
fig.show()

# 0 = No Fraude
# 1 = Fraude

## Selección de threshold por filas y columnas para eliminar valores missing

In [11]:
bf_series_null_columns = bf.isnull().sum().sort_values(ascending=False)
bf_series_null_rows = bf.isnull().sum(axis=1).sort_values(ascending=False)
print(bf_series_null_columns.shape, bf_series_null_rows.shape)

bf_null_columnas = pd.DataFrame(bf_series_null_columns, columns=['nulos_columnas'])     
bf_null_filas = pd.DataFrame(bf_series_null_rows, columns=['nulos_filas'])  
bf_null_filas['target'] = bf['fraud_bool'].copy()
bf_null_columnas['porcentaje_columnas'] = bf_null_columnas['nulos_columnas']/bf.shape[0]
bf_null_filas['porcentaje_filas']= bf_null_filas['nulos_filas']/bf.shape[1]
# # pd_null

(32,) (1000000,)


In [12]:
bf.shape

(1000000, 32)

In [13]:
bf_null_columnas

nulos_columnas  porcentaje_columnas
fraud_bool                                     0                  0.0
income                                         0                  0.0
device_fraud_count                             0                  0.0
device_distinct_emails_8w                      0                  0.0
keep_alive_session                             0                  0.0
device_os                                      0                  0.0
session_length_in_minutes                      0                  0.0
source                                         0                  0.0
foreign_request                                0                  0.0
proposed_credit_limit                          0                  0.0
has_other_cards                                0                  0.0
bank_months_count                              0                  0.0
phone_mobile_valid                             0                  0.0
phone_home_valid                               0                  0.0
housing_status                                 0                  0.0
email_is_free                                  0                  0.0
credit_risk_score                              0                  0.0
employment_status                              0                  0.0
date_of_birth_distinct_emails_4w               0                  0.0
bank_branch_count_8w                           0                  0.0
velocity_4w                                    0                  0.0
velocity_24h                                   0                  0.0
velocity_6h                                    0                  0.0
zip_count_4w                                   0                  0.0
payment_type                                   0                  0.0
intended_balcon_amount                         0                  0.0
days_since_request                             0                  0.0
customer_age                                   0                  0.0
current_address_months_count                   0                  0.0
prev_address_months_count                      0                  0.0
name_email_similarity                          0                  0.0
month                                          0                  0.0

Podemos observar que el número de valores nulos identificados en el dataset es de 0,por lo tanto no es necesario modificarlo para eliminar valores nulos.

In [14]:
threshold=0.9
list_vars_not_null = list(bf_null_columnas[bf_null_columnas['porcentaje_columnas']<threshold].index)
bf_filter_null = bf.loc[:, list_vars_not_null]
bf_filter_null.shape

(1000000, 32)

Podemos confirmar que no hay valores nulos.

In [15]:
bf_null_filas


nulos_filas  target  porcentaje_filas
0                 0       1               0.0
666657            0       0               0.0
666659            0       0               0.0
666660            0       0               0.0
666661            0       0               0.0
...             ...     ...               ...
333337            0       0               0.0
333338            0       0               0.0
333339            0       0               0.0
333340            0       0               0.0
999999            0       0               0.0

[1000000 rows x 3 columns]

# <font color = 'lightgreen'> INTERPRETACIÓN DE LOS VALORES NULOS

Hay algunas variables del dataframe que contienen valores nulos, generalmente se identifican como -1. Las variables que contienen estos valores nulos son: "prev_address_months_count","current_address_months_count","intended_balcon_amount", "bank_months_count", "session_length_in_minutes", "device_distinct_emails_8w". 

Para la interpretación de dichos valores hemos creado las variables "valores_nulos", donde observaremos el número total de NAs y su porcentaje respecto de la suma total de datos de cada variable:

In [28]:
valores_nulos1 = (bf["prev_address_months_count"] < 0).astype(int).sum()
valores_nulos2 = (bf["current_address_months_count"] < 0).astype(int).sum()
valores_nulos3 = (bf["intended_balcon_amount"] < 0).astype(int).sum()
valores_nulos4 = (bf["bank_months_count"] < 0).astype(int).sum()
valores_nulos5 = (bf["session_length_in_minutes"] < 0).astype(int).sum()
valores_nulos6 = (bf["device_distinct_emails_8w"] < 0).astype(int).sum()

print(f"La cantidad de valores nulos en 'prev_address_months_count' es: {valores_nulos1}")
print(f"La cantidad de valores nulos en 'current_address_months_count' es: {valores_nulos2}")
print(f"La cantidad de valores nulos en ' intended_balcon_amount' es: {valores_nulos3}")
print(f"La cantidad de valores nulos en 'bank_months_count' es: {valores_nulos4}")
print(f"La cantidad de valores nulos en 'session_length_in_minutes' es: {valores_nulos5}")
print(f"La cantidad de valores nulos en 'device_distinct_emails ' es: {valores_nulos6}")

La cantidad de valores nulos en 'prev_address_months_count' es: 712920
La cantidad de valores nulos en 'current_address_months_count' es: 4254
La cantidad de valores nulos en ' intended_balcon_amount' es: 742523
La cantidad de valores nulos en 'bank_months_count' es: 253635
La cantidad de valores nulos en 'session_length_in_minutes' es: 2015
La cantidad de valores nulos en 'device_distinct_emails ' es: 359


In [29]:
print(f"Esto supone que los valores nulos en prev_address_months_count representan un ",(valores_nulos1/1000000)*100,"%" )
print(f"Esto supone que los valores nulos en current_address_months_count representan un ",(valores_nulos2/1000000)*100,"%" )
print(f"Esto supone que los valores nulos en intended_balcon_amount representan un ",(valores_nulos3/1000000)*100,"%" )
print(f"Esto supone que los valores nulos en 'bank_months_count representan un ",(valores_nulos4/1000000)*100,"%" )
print(f"Esto supone que los valores nulos en session_length_in_minutes representan un ",(valores_nulos5/1000000)*100,"%" )
print(f"Esto supone que los valores nulos en device_distinct_emails representan un ",(valores_nulos6/1000000)*100,"%" )

Esto supone que los valores nulos en prev_address_months_count representan un  71.292 %
Esto supone que los valores nulos en current_address_months_count representan un  0.4254 %
Esto supone que los valores nulos en intended_balcon_amount representan un  74.2523 %
Esto supone que los valores nulos en 'bank_months_count representan un  25.3635 %
Esto supone que los valores nulos en session_length_in_minutes representan un  0.20149999999999998 %
Esto supone que los valores nulos en device_distinct_emails representan un  0.0359 %


# Tipos: Variables categoricas y numericas

In [30]:
list_cat_vars = list(new_bf.select_dtypes(include=['category']).columns)


Creamos una lista que contenga solo las variables categóricas del dataset.

In [31]:
bf_filter_null[list_cat_vars] = bf_filter_null[list_cat_vars].astype("category")
bf_filter_null[list_cat_vars].head()

fraud_bool payment_type employment_status email_is_free housing_status  \
0          1           AA                CA             0             BA   
1          1           AB                CA             1             BA   
2          1           AC                CB             1             BA   
3          1           AB                CA             1             BA   
4          1           AB                CA             1             BA   

  phone_home_valid phone_mobile_valid has_other_cards foreign_request  \
0                1                  0               0               0   
1                0                  0               0               0   
2                0                  1               0               0   
3                0                  1               1               0   
4                1                  0               0               0   

     source  device_os keep_alive_session  
0  INTERNET    windows                  0  
1  INTERNET    windows                  0  
2  INTERNET      other                  0  
3  INTERNET      linux                  0  
4  INTERNET  macintosh                  1

In [32]:
list_cat_vars


['fraud_bool',
 'payment_type',
 'employment_status',
 'email_is_free',
 'housing_status',
 'phone_home_valid',
 'phone_mobile_valid',
 'has_other_cards',
 'foreign_request',
 'source',
 'device_os',
 'keep_alive_session']

Observamos todas las variables que son de tipo categórica creadas en la lista (list_cat_vars)

In [33]:
bf_filter_null['device_os'].value_counts()


other        342728
linux        332712
windows      263506
macintosh     53826
x11            7228
Name: device_os, dtype: int64

Observamos la frecuencia de valores únicos dentro de "device_os", es decir las veces que se repite un mismo dato dentro de la variable que en este caso se divide en other, la cual tiene el mayor número de repeticiones,linux,windows,macintosh y x11

In [34]:
bf_filter_null['customer_age'].value_counts()


30    311433
20    245855
40    238712
50    140353
60     34770
10     20987
70      6517
80      1297
90        76
Name: customer_age, dtype: int64

Realizamos el mismo procedimiento que en la línea de código anterior, esta vez con la variable categórica "customer age"

In [35]:
bf_filter_null[list_cat_vars].dtypes

fraud_bool            category
payment_type          category
employment_status     category
email_is_free         category
housing_status        category
phone_home_valid      category
phone_mobile_valid    category
has_other_cards       category
foreign_request       category
source                category
device_os             category
keep_alive_session    category
dtype: object

In [36]:
print(bf_filter_null["credit_risk_score"].value_counts().count())
bf_filter_null["credit_risk_score"].value_counts(normalize=True)

551


 113    0.006791
 116    0.006784
 110    0.006774
 115    0.006734
 117    0.006717
 114    0.006703
 109    0.006676
 105    0.006674
 112    0.006653
 111    0.006624
 108    0.006618
 107    0.006611
 103    0.006607
 118    0.006590
 100    0.006559
 106    0.006547
 104    0.006545
 102    0.006500
 120    0.006488
 101    0.006447
 122    0.006419
 119    0.006410
 98     0.006402
 121    0.006375
 99     0.006345
 123    0.006331
 96     0.006291
 97     0.006233
 95     0.006229
 94     0.006183
 125    0.006090
 124    0.006070
 126    0.006042
 92     0.005959
 90     0.005958
 91     0.005932
 93     0.005915
 127    0.005874
 129    0.005871
 88     0.005851
 89     0.005838
 128    0.005796
 130    0.005767
 87     0.005762
 86     0.005750
 85     0.005666
 83     0.005653
 84     0.005575
 131    0.005517
 81     0.005510
 132    0.005498
 82     0.005484
 133    0.005415
 80     0.005341
 134    0.005276
 78     0.005223
 77     0.005173
 136    0.005148
 135    0.0051

En esta línea de código primeramente imprimimos el total de valores únicos dentro de "credit_risk_score", en este caso son 551 ,más adelante en la columna izquierda aparecen cada uno de los 551 valores únicos acompañado en la columna derecha con la proporción de cada valor, es decir el valor 113 dentro de credit_risk_score tiene una proporción de 0.6%.

## Preprocesamiento inicial de algunas variables

In [37]:
bf_filter_null.head()


fraud_bool  income  device_fraud_count  device_distinct_emails_8w  \
0          1     0.9                   0                          1   
1          1     0.9                   0                          1   
2          1     0.9                   0                          1   
3          1     0.9                   0                          1   
4          1     0.9                   0                          1   

  keep_alive_session  device_os  session_length_in_minutes    source  \
0                  0    windows                   3.888115  INTERNET   
1                  0    windows                  31.798819  INTERNET   
2                  0      other                   4.728705  INTERNET   
3                  0      linux                   2.047904  INTERNET   
4                  1  macintosh                   3.775225  INTERNET   

  foreign_request  proposed_credit_limit has_other_cards  bank_months_count  \
0               0                  500.0               0                 24   
1               0                 1500.0               0                 15   
2               0                  200.0               0                 -1   
3               0                  200.0               1                 31   
4               0                 1500.0               0                 31   

  phone_mobile_valid phone_home_valid housing_status email_is_free  \
0                  0                1             BA             0   
1                  0                0             BA             1   
2                  1                0             BA             1   
3                  1                0             BA             1   
4                  0                1             BA             1   

   credit_risk_score employment_status  date_of_birth_distinct_emails_4w  \
0                185                CA                                 6   
1                259                CA                                 3   
2                177                CB                                14   
3                110                CA                                 6   
4                295                CA                                 2   

   bank_branch_count_8w  velocity_4w  velocity_24h   velocity_6h  \
0                     1  3863.647740   3134.319630  10650.765523   
1                   718  3124.298166   2670.918292    534.047319   
2                     1  3159.590679   2893.621498   4048.534263   
3                  1921  3022.261812   4054.908412   3457.064063   
4                  1990  3087.670952   2728.237159   5020.341679   

   zip_count_4w payment_type  intended_balcon_amount  days_since_request  \
0           769           AA               -1.331345            0.020925   
1           366           AB               -0.816224            0.005418   
2           870           AC               -0.755728            3.108549   
3           810           AB               -1.205124            0.019079   
4           890           AB               -0.773276            0.004441   

   customer_age  current_address_months_count  prev_address_months_count  \
0            50                            88                         -1   
1            50                           144                         -1   
2            40                           132                         -1   
3            50                            22                         -1   
4            50                           218                         -1   

   name_email_similarity  month  
0               0.166828      7  
1               0.296286      7  
2               0.044985      7  
3               0.159511      7  
4               0.596414      7

In [38]:
bf_filter_null.shape


(1000000, 32)

En este caso, no ha variado nada el dataframe original del final, ya que no hay valores NAs (hay valores missings pero cada variable tiene un valor diferente), por lo que en el segundo notebook hemos utilizado el dataset 'Base.csv' sin tocar nada. Tampoco hemos ejecutado esta última línea de código ya que vemos irrelevante usar este dataframe o el otro.

In [39]:
#bf_filter_null.drop("earliest_cr_line",axis=1).to_csv("./data_preprocessing/pd_data_initial_preprocessing.csv")

En el segundo notebook cambiaremos ciertas variables de formato para poder continuar con el proyecto.

# <font color = 'lightgreen'> CONCLUSIÓN

Como conclusión de esta parte del trabajo hemos tratado todas las variables, definiendo y diferenciando tanto las númericas como categóricas y flotantes, hemos identificado la cantidad de variables nulas donde por ejemplo la variable prev_address_months_count tiene más de un 70% de valores nulos. Al ser los valores nulos valores negativos, solo los identificamos y no procedemos a eliminarlos, aun así tratamos los valores nulos donde observamos que en este caso en el csv base no se encuentra ninguno. Además de esto creamos la lista de variables categóricas para tratarlas.